File: frontend/src/api/tak.ts
import apiClient from './client';
import { isDemoMode } from './mockClient';

// --- Types ---

export interface TAKConnection {
  id: number;
  name: string;
  host: string;
  port: number;
  protocol: 'tcp' | 'ssl' | 'http';
  api_port: number;
  use_tls: boolean;
  cert_file: string | null;
  api_token: string | null;
  unit_id: number | null;
  cot_types_filter: string[] | null;
  channel_filter: string | null;
  is_active: boolean;
  connection_status: 'connected' | 'disconnected' | 'error' | 'polling';
  last_connected: string | null;
  last_error: string | null;
  created_by: number | null;
  created_at: string;
  updated_at: string | null;
}

export interface TAKConnectionCreate {
  name: string;
  host: string;
  port?: number;
  protocol?: 'tcp' | 'ssl' | 'http';
  api_port?: number;
  use_tls?: boolean;
  cert_file?: string;
  api_token?: string;
  unit_id?: number;
  cot_types_filter?: string[];
  channel_filter?: string;
  is_active?: boolean;
}

export interface TAKConnectionUpdate {
  name?: string;
  host?: string;
  port?: number;
  protocol?: 'tcp' | 'ssl' | 'http';
  api_port?: number;
  use_tls?: boolean;
  cert_file?: string;
  api_token?: string;
  unit_id?: number;
  cot_types_filter?: string[];
  channel_filter?: string;
  is_active?: boolean;
}

export interface TAKConnectionStatus {
  connection_id: number;
  name: string;
  status: string;
  connected_at: string | null;
  last_poll: string | null;
  last_error: string | null;
  messages_received: number;
}

export interface TAKTestResult {
  success: boolean;
  status_code: number | null;
  latency_ms: number | null;
  server_info: string | null;
  message: string;
}

export interface TAKMessage {
  id: number;
  uid: string | null;
  callsign: string | null;
  channel: string | null;
  content_preview: string | null;
  confidence: number;
  status: string;
  received_at: string | null;
}

export interface COTIngestResponse {
  success: boolean;
  message: string;
  parsed: Record<string, unknown> | null;
  errors: string[] | null;
}

// --- Demo mock data for TAK ---

const DEMO_TAK_CONNECTION: TAKConnection = {
  id: 1,
  name: 'TAK Server Alpha',
  host: 'tak-alpha.keystone.usmc.mil',
  port: 8089,
  protocol: 'ssl',
  api_port: 8443,
  use_tls: true,
  cert_file: '/certs/tak-alpha.pem',
  api_token: null,
  unit_id: 4,
  cot_types_filter: ['a-f-G', 'b-m-p-s-m'],
  channel_filter: null,
  is_active: true,
  connection_status: 'connected',
  last_connected: new Date().toISOString(),
  last_error: null,
  created_by: 1,
  created_at: '2026-02-15T00:00:00Z',
  updated_at: '2026-03-01T12:00:00Z',
};

const DEMO_TAK_MESSAGES: TAKMessage[] = [
  { id: 1, uid: 'ALPHA-6', callsign: 'ALPHA-6', channel: 'BN TAC 1', content_preview: 'Position update: Grid 38SMB 4823 1567', confidence: 0.95, status: 'PARSED', received_at: new Date(Date.now() - 300000).toISOString() },
  { id: 2, uid: 'BRAVO-6', callsign: 'BRAVO-6', channel: 'BN TAC 1', content_preview: 'LOGPAC request: CL III emergency resupply', confidence: 0.88, status: 'PARSED', received_at: new Date(Date.now() - 600000).toISOString() },
  { id: 3, uid: 'FST-11', callsign: 'STEEL RAIN', channel: 'FIRES NET', content_preview: 'Fire mission complete. BDA pending.', confidence: 0.92, status: 'PARSED', received_at: new Date(Date.now() - 900000).toISOString() },
];

const mockDelay = (ms = 200 + Math.random() * 200): Promise<void> =>
  new Promise((r) => setTimeout(r, ms));

// --- API Functions ---

export async function getTAKConnections(): Promise<TAKConnection[]> {
  if (isDemoMode) { await mockDelay(); return [DEMO_TAK_CONNECTION]; }
  const response = await apiClient.get<TAKConnection[]>('/tak/connections');
  return response.data;
}

export async function createConnection(data: TAKConnectionCreate): Promise<TAKConnection> {
  if (isDemoMode) { await mockDelay(); return { ...DEMO_TAK_CONNECTION, ...data, id: Date.now() }; }
  const response = await apiClient.post<TAKConnection>('/tak/connections', data);
  return response.data;
}

export async function updateConnection(
  id: number,
  data: TAKConnectionUpdate,
): Promise<TAKConnection> {
  if (isDemoMode) { await mockDelay(); return { ...DEMO_TAK_CONNECTION, ...data, id }; }
  const response = await apiClient.put<TAKConnection>(`/tak/connections/${id}`, data);
  return response.data;
}

export async function deleteConnection(id: number): Promise<void> {
  if (isDemoMode) { await mockDelay(); void id; return; }
  await apiClient.delete(`/tak/connections/${id}`);
}

export async function testConnection(id: number): Promise<TAKTestResult> {
  if (isDemoMode) { await mockDelay(); void id; return { success: true, status_code: 200, latency_ms: 45, server_info: 'TAK Server 4.8.0 (Demo)', message: 'Connection successful' }; }
  const response = await apiClient.post<TAKTestResult>(`/tak/connections/${id}/test`);
  return response.data;
}

export async function startPolling(id: number): Promise<{ status: string; message: string }> {
  if (isDemoMode) { await mockDelay(); void id; return { status: 'polling', message: 'Polling started (demo)' }; }
  const response = await apiClient.post<{ status: string; message: string }>(
    `/tak/connections/${id}/start`,
  );
  return response.data;
}

export async function stopPolling(id: number): Promise<{ status: string; message: string }> {
  if (isDemoMode) { await mockDelay(); void id; return { status: 'disconnected', message: 'Polling stopped (demo)' }; }
  const response = await apiClient.post<{ status: string; message: string }>(
    `/tak/connections/${id}/stop`,
  );
  return response.data;
}

export async function getConnectionStatus(id: number): Promise<TAKConnectionStatus> {
  if (isDemoMode) { await mockDelay(); return { connection_id: id, name: DEMO_TAK_CONNECTION.name, status: 'connected', connected_at: DEMO_TAK_CONNECTION.last_connected, last_poll: new Date().toISOString(), last_error: null, messages_received: 47 }; }
  const response = await apiClient.get<TAKConnectionStatus>(`/tak/connections/${id}/status`);
  return response.data;
}

export async function getRecentMessages(id: number, limit = 20): Promise<TAKMessage[]> {
  if (isDemoMode) { await mockDelay(); void id; return DEMO_TAK_MESSAGES.slice(0, limit); }
  const response = await apiClient.get<TAKMessage[]>(`/tak/connections/${id}/messages`, {
    params: { limit },
  });
  return response.data;
}

export async function submitCOTMessage(
  xmlContent: string,
  connectionId?: number,
): Promise<COTIngestResponse> {
  if (isDemoMode) { await mockDelay(); void connectionId; return { success: true, message: 'CoT message ingested (demo)', parsed: { type: 'a-f-G', uid: 'DEMO-' + Date.now(), content: xmlContent.substring(0, 100) }, errors: null }; }
  const response = await apiClient.post<COTIngestResponse>('/tak/ingest/cot', {
    xml_content: xmlContent,
    connection_id: connectionId,
  });
  return response.data;
}

File: backend/app/api/tak.py
"""TAK (Team Awareness Kit) server management API endpoints."""

import ipaddress
import socket
from typing import List, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.core.auth import get_current_user
from app.core.exceptions import BadRequestError, NotFoundError
from app.core.permissions import require_role
from app.database import get_db
from app.ingestion.tak_connector import tak_connector
from app.ingestion.tak_parser import parse_cot_message
from app.models.tak_connection import ConnectionStatus, TAKConnection
from app.models.user import Role, User
from app.schemas.tak import (
    COTIngestRequest,
    COTIngestResponse,
    COTMessage,
    TAKConnectionCreate,
    TAKConnectionResponse,
    TAKConnectionStatus,
    TAKConnectionUpdate,
    TAKTestResult,
)

router = APIRouter()

# Private IP networks that should be blocked in production (SSRF protection)
_PRIVATE_NETWORKS = [
    ipaddress.ip_network("127.0.0.0/8"),
    ipaddress.ip_network("10.0.0.0/8"),
    ipaddress.ip_network("172.16.0.0/12"),
    ipaddress.ip_network("192.168.0.0/16"),
    ipaddress.ip_network("169.254.0.0/16"),
    ipaddress.ip_network("::1/128"),
]

# Cloud metadata endpoints that must always be blocked
_BLOCKED_HOSTS = {
    "metadata.google.internal",
    "metadata.google.com",
    "169.254.169.254",
}


def _validate_tak_host(host: str) -> None:
    """Validate that a TAK host is not a private/internal address.

    Blocks SSRF attacks by rejecting private IP ranges and cloud
    metadata endpoints. Controlled by ALLOW_PRIVATE_TAK_HOSTS setting.

    Raises:
        BadRequestError: If the host resolves to a blocked address.
    """
    # Always block cloud metadata endpoints
    if host.lower() in _BLOCKED_HOSTS:
        raise BadRequestError(f"Host '{host}' is a blocked cloud metadata endpoint.")

    # If private hosts are allowed (dev mode), skip further checks
    if settings.ALLOW_PRIVATE_TAK_HOSTS:
        return

    # Resolve hostname to IP and check against private ranges
    try:
        addr_infos = socket.getaddrinfo(host, None)
        for addr_info in addr_infos:
            ip = ipaddress.ip_address(addr_info[4][0])
            for network in _PRIVATE_NETWORKS:
                if ip in network:
                    raise BadRequestError(
                        f"Host '{host}' resolves to private IP {ip}. "
                        "Private TAK hosts are not allowed in production. "
                        "Set ALLOW_PRIVATE_TAK_HOSTS=true for development."
                    )
    except socket.gaierror:
        # Cannot resolve - let the connection attempt fail naturally
        pass


@router.get("/connections", response_model=List[TAKConnectionResponse])
async def list_tak_connections(
    is_active: Optional[bool] = Query(None),
    limit: int = Query(50, le=200),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """List all configured TAK server connections."""
    query = select(TAKConnection)

    if is_active is not None:
        query = query.where(TAKConnection.is_active == is_active)

    query = query.order_by(TAKConnection.created_at.desc()).offset(offset).limit(limit)
    result = await db.execute(query)
    connections = result.scalars().all()

    return [TAKConnectionResponse.model_validate(conn) for conn in connections]


@router.post("/connections", response_model=TAKConnectionResponse)
async def create_tak_connection(
    data: TAKConnectionCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_role([Role.ADMIN, Role.COMMANDER, Role.S4])),
):
    """Create a new TAK server connection.

    Requires ADMIN, COMMANDER, or S4 role.
    """
    _validate_tak_host(data.host)

    connection = TAKConnection(
        name=data.name,
        host=data.host,
        port=data.port,
        protocol=data.protocol,
        api_port=data.api_port,
        use_tls=data.use_tls,
        verify_tls=data.verify_tls,
        cert_file=data.cert_file,
        api_token=data.api_token,
        unit_id=data.unit_id,
        cot_types_filter=data.cot_types_filter,
        channel_filter=data.channel_filter,
        is_active=data.is_active,
        connection_status=ConnectionStatus.DISCONNECTED,
        created_by=current_user.id,
    )

    db.add(connection)
    await db.flush()
    await db.refresh(connection)

    return TAKConnectionResponse.model_validate(connection)


@router.put("/connections/{connection_id}", response_model=TAKConnectionResponse)
async def update_tak_connection(
    connection_id: int,
    data: TAKConnectionUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_role([Role.ADMIN, Role.COMMANDER, Role.S4])),
):
    """Update an existing TAK server connection.

    Requires ADMIN, COMMANDER, or S4 role.
    """
    result = await db.execute(
        select(TAKConnection).where(TAKConnection.id == connection_id)
    )
    connection = result.scalar_one_or_none()
    if not connection:
        raise NotFoundError("TAK Connection", connection_id)

    update_data = data.model_dump(exclude_unset=True)

    # Validate host if it's being changed
    if "host" in update_data:
        _validate_tak_host(update_data["host"])

    for field, value in update_data.items():
        setattr(connection, field, value)

    await db.flush()
    await db.refresh(connection)

    return TAKConnectionResponse.model_validate(connection)


@router.delete("/connections/{connection_id}")
async def delete_tak_connection(
    connection_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_role([Role.ADMIN])),
):
    """Delete a TAK server connection.

    Requires ADMIN role. Disconnects any active connection before deleting.
    """
    result = await db.execute(
        select(TAKConnection).where(TAKConnection.id == connection_id)
    )
    connection = result.scalar_one_or_none()
    if not connection:
        raise NotFoundError("TAK Connection", connection_id)

    # Disconnect if active
    await tak_connector.disconnect(connection_id)

    await db.delete(connection)
    await db.flush()

    return {
        "id": connection_id,
        "message": "TAK connection deleted",
    }


@router.post("/connections/{connection_id}/test", response_model=TAKTestResult)
async def test_tak_connection(
    connection_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Test connectivity to a TAK server.

    Attempts a connection to the configured TAK server and reports
    the result including latency and server version.
    """
    result = await db.execute(
        select(TAKConnection).where(TAKConnection.id == connection_id)
    )
    connection = result.scalar_one_or_none()
    if not connection:
        raise NotFoundError("TAK Connection", connection_id)

    test_result = await tak_connector.test_connection(connection)
    return TAKTestResult(**test_result)


@router.post("/connections/{connection_id}/start")
async def start_tak_polling(
    connection_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_role([Role.ADMIN, Role.COMMANDER, Role.S4])),
):
    """Start polling a TAK server for CoT messages.

    Connects to the TAK server and enqueues a periodic polling task.
    Requires ADMIN, COMMANDER, or S4 role.
    """
    result = await db.execute(
        select(TAKConnection).where(TAKConnection.id == connection_id)
    )
    connection = result.scalar_one_or_none()
    if not connection:
        raise NotFoundError("TAK Connection", connection_id)

    if not connection.is_active:
        raise BadRequestError(
            "Cannot start polling on an inactive connection. "
            "Enable the connection first."
        )

    # Attempt to connect
    connected = await tak_connector.connect(connection)

    if connected:
        # Update connection status in DB
        connection.connection_status = ConnectionStatus.CONNECTED
        await db.flush()

        # Enqueue initial poll task
        try:
            from app.tasks.poll_tak import poll_tak_server

            poll_tak_server.delay(connection_id)
        except Exception:
            pass  # Celery may not be running in dev

        return {
            "connection_id": connection_id,
            "status": "connected",
            "message": f"Polling started for {connection.name}",
        }
    else:
        status = tak_connector.get_connection_status(connection_id)
        connection.connection_status = ConnectionStatus.ERROR
        connection.last_error = status.get("last_error", "Connection failed")
        await db.flush()

        raise BadRequestError(
            f"Failed to connect: {status.get('last_error', 'Unknown error')}"
        )


@router.post("/connections/{connection_id}/stop")
async def stop_tak_polling(
    connection_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_role([Role.ADMIN, Role.COMMANDER, Role.S4])),
):
    """Stop polling a TAK server.

    Disconnects from the TAK server and stops the periodic polling task.
    Requires ADMIN, COMMANDER, or S4 role.
    """
    result = await db.execute(
        select(TAKConnection).where(TAKConnection.id == connection_id)
    )
    connection = result.scalar_one_or_none()
    if not connection:
        raise NotFoundError("TAK Connection", connection_id)

    await tak_connector.disconnect(connection_id)

    connection.connection_status = ConnectionStatus.DISCONNECTED
    await db.flush()

    return {
        "connection_id": connection_id,
        "status": "disconnected",
        "message": f"Polling stopped for {connection.name}",
    }


@router.get(
    "/connections/{connection_id}/status",
    response_model=TAKConnectionStatus,
)
async def get_tak_connection_status(
    connection_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get the current status of a TAK connection.

    Returns connection state, message counts, last poll time, and any errors.
    """
    result = await db.execute(
        select(TAKConnection).where(TAKConnection.id == connection_id)
    )
    connection = result.scalar_one_or_none()
    if not connection:
        raise NotFoundError("TAK Connection", connection_id)

    runtime_status = tak_connector.get_connection_status(connection_id)

    return TAKConnectionStatus(
        connection_id=connection_id,
        name=connection.name,
        status=runtime_status.get("status", connection.connection_status.value),
        connected_at=runtime_status.get("connected_at"),
        last_poll=runtime_status.get("last_poll"),
        last_error=runtime_status.get("last_error") or connection.last_error,
        messages_received=runtime_status.get("messages_received", 0),
    )


@router.get("/connections/{connection_id}/messages")
async def get_tak_messages(
    connection_id: int,
    limit: int = Query(20, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get recent messages received from a TAK connection.

    Returns parsed CoT messages stored as raw data records from
    this TAK connection.
    """
    result = await db.execute(
        select(TAKConnection).where(TAKConnection.id == connection_id)
    )
    connection = result.scalar_one_or_none()
    if not connection:
        raise NotFoundError("TAK Connection", connection_id)

    from app.models.raw_data import RawData

    # Query raw data records associated with this TAK connection
    messages_result = await db.execute(
        select(RawData)
        .where(
            RawData.file_name.like("tak_cot_%"),
            RawData.channel_name == connection.channel_filter
            if connection.channel_filter
            else True,
        )
        .order_by(RawData.ingested_at.desc())
        .limit(limit)
    )
    messages = messages_result.scalars().all()

    return [
        {
            "id": msg.id,
            "uid": msg.file_name.replace("tak_cot_", "") if msg.file_name else None,
            "callsign": msg.sender,
            "channel": msg.channel_name,
            "content_preview": (
                msg.original_content[:200] if msg.original_content else None
            ),
            "confidence": msg.confidence_score,
            "status": msg.parse_status.value,
            "received_at": (msg.ingested_at.isoformat() if msg.ingested_at else None),
        }
        for msg in messages
    ]


@router.post("/ingest/cot", response_model=COTIngestResponse)
async def ingest_cot_message(
    data: COTIngestRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Manually submit a CoT XML message for parsing.

    Useful for testing CoT parsing without a live TAK server connection.
    Parses the XML and optionally stores the result.
    """
    parsed = parse_cot_message(data.xml_content)

    if parsed is None:
        return COTIngestResponse(
            success=False,
            message="Failed to parse CoT XML. Ensure valid CoT event format.",
            errors=["Invalid or malformed CoT XML"],
        )

    # Store as raw data for audit
    from app.models.raw_data import ParseStatus, RawData, SourceType

    raw = RawData(
        source_type=SourceType.MANUAL,
        original_content=data.xml_content,
        channel_name=parsed.get("group_name"),
        sender=parsed.get("callsign"),
        file_name=f"tak_cot_{parsed.get('uid', 'manual')}",
        parse_status=ParseStatus.PARSED,
        confidence_score=1.0,
        reviewed_by=current_user.id,
    )
    db.add(raw)
    await db.flush()

    # Build response
    cot_message = COTMessage(
        uid=parsed.get("uid", ""),
        cot_type=parsed.get("cot_type", ""),
        category=parsed.get("category", "UNKNOWN"),
        event_time=parsed.get("event_time"),
        start_time=parsed.get("start_time"),
        stale_time=parsed.get("stale_time"),
        how=parsed.get("how"),
        callsign=parsed.get("callsign"),
        group_name=parsed.get("group_name"),
        group_role=parsed.get("group_role"),
        remarks=parsed.get("remarks"),
    )

    # Add position if present
    if parsed.get("position"):
        from app.schemas.tak import COTPosition

        cot_message.position = COTPosition(**parsed["position"])

    return COTIngestResponse(
        success=True,
        message=f"CoT message parsed successfully. Category: {cot_message.category}",
        parsed=cot_message,
    )
